File: python_pid/pid.py
"""PID controller implementation.

This module provides a PID controller class using the incremental
(velocity) form, which provides intrinsic integrator anti-windup and
bumpless transfer behavior.
"""

from .anti_windup import anti_windup
from .measurement_filter import MeasurementFilter


def pid_update(
    r,
    yf,
    dyf,
    kp,
    ki,
    kd,
    umin,
    umax,
    u0,
    b,
    u_old,
    up_old,
    ud_old,
    uff_old,
    uff=0.0,
    uman=0.0,
    utrack=0.0,
    Tx=1.0,
    track=False,
    auto=True,
    windup=None
):
    """PID control signal update.

    This is a standalone function implementing the incremental PID
    algorithm.

    Args:
        r: Reference (setpoint) signal
        yf: Filtered measurement
        dyf: Filtered derivative of measurement
        kp: Proportional gain
        ki: Integral gain
        kd: Derivative gain
        umin: Minimum control signal
        umax: Maximum control signal
        u0: Bias term for P or PD control
        b: Setpoint weight for proportional term
        u_old: Previous control signal
        up_old: Previous proportional component
        ud_old: Previous derivative component
        uff_old: Previous feedforward signal
        uff: Feedforward control signal (default: 0.0)
        uman: Manual mode control signal (default: 0.0)
        utrack: Tracking signal for bumpless transfer (default: 0.0)
        Tx: Execution period normalized (default: 1.0)
        track: Tracking mode flag (default: False)
        auto: Automatic mode flag (default: True)
        windup: Windup status ("upper", "lower", "both", or None)

    Returns:
        tuple: (u, u_old, up_old, ud_old, uff_old, b)
               Control signal and updated state values
    """
    if auto:
        # Reset state if using P or PD control (ki == 0)
        if ki == 0:
            u_old = u0  # Bias term if P or PD control
            up_old = 0.0
            ud_old = 0.0
            uff_old = 0.0
            b = 1.0

        # Tracking mode for bumpless transfer
        if track:
            u_old = utrack
            up_old = 0.0
            ud_old = 0.0
            uff_old = 0.0

        # Control signal increments
        Dup = kp * (b * r - yf) - up_old
        Dui = ki * (r - yf) * Tx
        Dui = anti_windup(Dui, windup)
        Dud = (-kd * dyf - ud_old) / Tx
        Duff = uff - uff_old

        # Add control signal increment
        Du = Dup + Dui + Dud + Duff
        u = u_old + Du
    else:
        # Manual control signal
        u = uman

    # Saturate control signal
    u = max(min(u, umax), umin)

    # Compute new state values (to be used in next iteration)
    u_old = u  # Store saturated control signal
    up_old = kp * (b * r - yf)
    ud_old = -kd * dyf
    uff_old = uff

    return u, u_old, up_old, ud_old, uff_old, b


class PIDController:
    """PID controller using incremental (velocity) form.

    The incremental form computes control signal increments rather than
    absolute values, providing natural anti-windup and bumpless transfer
    capabilities.
    """

    def __init__(
        self,
        kp,
        ki,
        kd,
        TfTs=10.0,
        umin=-float('inf'),
        umax=float('inf'),
        u0=0.0,
        b=1.0
    ):
        """Initialize the PID controller.

        Args:
            kp: Proportional gain
            ki: Integral gain
            kd: Derivative gain
            TfTs: Filter time constant as multiple of nominal sample
                  time (default: 10.0)
            umin: Minimum control signal (default: -inf)
            umax: Maximum control signal (default: +inf)
            u0: Bias term for P or PD control (default: 0.0)
            b: Setpoint weight for proportional term (default: 1.0)
        """
        # Controller parameters
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.umin = umin
        self.umax = umax
        self.u0 = u0
        self.b = b

        # Signal states
        self.u_old = 0.0
        self.up_old = 0.0
        self.ud_old = 0.0
        self.uff_old = 0.0

        # Measurement filter
        self.filter = MeasurementFilter(TfTs=TfTs)

    def __call__(
        self,
        r,
        y,
        uff=0.0,
        uman=0.0,
        utrack=0.0,
        Tx=1.0,
        track=False,
        auto=True,
        windup=None
    ):
        """Compute the PID control signal.

        Args:
            r: Reference (setpoint) signal
            y: Process measurement
            uff: Feedforward control signal (default: 0.0)
            uman: Manual mode control signal (default: 0.0)
            utrack: Tracking signal for bumpless transfer (default: 0.0)
            Tx: Execution period normalized (default: 1.0)
            track: Tracking mode flag (default: False)
            auto: Automatic mode flag (default: True)
            windup: Windup status ("upper", "lower", "both", or None)

        Returns:
            float: Control signal u
        """
        # Filter updates
        yf, dyf = self.filter(y, Tx)

        # Call standalone PID update function
        u, self.u_old, self.up_old, self.ud_old, self.uff_old, self.b = (
            pid_update(
                r, yf, dyf,
                self.kp, self.ki, self.kd,
                self.umin, self.umax, self.u0, self.b,
                self.u_old, self.up_old, self.ud_old, self.uff_old,
                uff, uman, utrack, Tx, track, auto, windup
            )
        )

        return u

    def reset(self):
        """Reset the controller state."""
        self.u_old = 0.0
        self.up_old = 0.0
        self.ud_old = 0.0
        self.uff_old = 0.0
        self.filter.reset()

File: cpp_pid/pid.cpp
/**
 * @file pid.cpp
 * @brief Implementation of PID controller
 */

#include "pid.h"
#include <algorithm>

PIDController::PIDController(
    double kp,
    double ki,
    double kd,
    double TfTs,
    double umin,
    double umax,
    double u0,
    double b)
    : kp_(kp),
      ki_(ki),
      kd_(kd),
      umin_(umin),
      umax_(umax),
      u0_(u0),
      b_(b),
      u_old_(0.0),
      up_old_(0.0),
      ud_old_(0.0),
      uff_old_(0.0),
      filter_(TfTs) {}

double PIDController::operator()(
    double r,
    double y,
    double uff,
    double uman,
    double utrack,
    double Tx,
    bool track,
    bool auto_mode,
    WindupMode windup) {

    // Filter updates
    FilterOutput filtered = filter_(y, Tx);
    double yf = filtered.yf;
    double dyf = filtered.dyf;

    double u;

    if (auto_mode) {
        // Reset state if using P or PD control (ki == 0)
        if (ki_ == 0.0) {
            u_old_ = u0_;  // Bias term if P or PD control
            up_old_ = 0.0;
            ud_old_ = 0.0;
            uff_old_ = 0.0;
            b_ = 1.0;
        }

        // Tracking mode for bumpless transfer
        if (track) {
            u_old_ = utrack;
            up_old_ = 0.0;
            ud_old_ = 0.0;
            uff_old_ = 0.0;
        }

        // Control signal increments
        double Dup = kp_ * (b_ * r - yf) - up_old_;
        double Dui = ki_ * (r - yf) * Tx;
        Dui = anti_windup(Dui, windup);
        double Dud = (-kd_ * dyf - ud_old_) / Tx;
        double Duff = uff - uff_old_;

        // Add control signal increment
        double Du = Dup + Dui + Dud + Duff;
        u = u_old_ + Du;
    } else {
        // Manual control signal
        u = uman;
    }

    // Saturate control signal
    u = std::max(std::min(u, umax_), umin_);

    // Update old signal states
    u_old_ = u;
    up_old_ = kp_ * (b_ * r - yf);
    ud_old_ = -kd_ * dyf;
    uff_old_ = uff;

    return u;
}

void PIDController::reset() {
    u_old_ = 0.0;
    up_old_ = 0.0;
    ud_old_ = 0.0;
    uff_old_ = 0.0;
    filter_.reset();
}
